Add yanking guidance to the readme faq (#104800)
JuliaRegistries/General: README.md

### Who can approve an early merge?

Any registry maintainer may merge a package registration PR early.
Early merges should still be discussed on the [#pkg-registration](https://julialang.slack.com/archives/C6M4DQA5P) slack channel before merging.
Registry maintainers are discouraged from merging packages they directly *or* indirectly maintain as mentioned in [CONTRIBUTING.md](https://github.com/JuliaRegistries/General/blob/master/CONTRIBUTING.md#other-ways-to-help).

### When is yanking a release appropriate?

Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release
entry in `Versions.toml` ([example](https://github.com/JuliaRegistries/General/pull/94858/files)). This
tells Pkg to no longer resolve those versions (e.g. when running `Pkg.add()` or `Pkg.resolve()`). However,
to maintain perfect reproducibility of environments specified with a full manifest, yanked versions that
are specifically listed in a manifest will continue to instantiate when running `Pkg.instantiate()` and

> yanked versions cannot be re-registered in General with different source code.

Reviewer: There has been discussion recently on slack (cc @mbauman) with the whole xz kerfuffle about no longer instantiating yanked versions either, in order to protect users from a vulnerability they may not even know about. This is also consistent with the next paragraph, stating the primary purpose of yanking to be "to make a release impossible to install". 

We should decide on a direction this should move in before writing down a behavioral guarantee for instantiation.

Author: For "impossible to install" we have deletion from the registry.

Reviewer: Deleting a version has a higher maintenance overhead for registry maintainers, since people then have to actively check the past git history for previously conflicting registrations. I don't think any previously registered version should ever be deleted from the registry, because it makes it much harder than necessary to keep track of changes to the registry.

The primary reason yanking exists is to make a release impossible to install. This is intended to be used for circumstances when having this
release installed poses a risk to security or safety. It is not intended for avoiding people installing releases with bugs, for that use a
patch release -- potentially after doing nothing more than `git revert`  in the package repo. All patch releases prior to the most recent have known bugs by definition.

There is however, a special category of bugged releases that can not be resolved by having a patch release. These also need to be resolved by
yanking. That special category is when the compat bounds have been set too wide. i.e. say `v2.10.0` was released using a feature not on julia
`v1.6` but the compat entry for julia was not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would
not solve the issue as on julia v1.6 Pkg would still resolve the broken `v2.10.0`, and as a minor bump, reverting the code changes would not

> be valid in a patch bump.

Reviewer: Could this sort of too-wide compat be resolved by making a PR to the registry with the correct bounds instead of yanking? After all, the release is (a priori) good, just not for all originally specified versions.

than the package erroring on load), then it may be yanked. Releasing an ordinary patch release—potentially 
reusing the exact same source code as a previous release without the major bug—is a faster, easier to 
deploy, and less disruptive way to prevent most users from installing a specific version. If you seek to
yank a very broken release, you should typically also release a patch release.

There is however, a special category of bugged releases that can not be resolved by having a patch release.
These also may to be resolved by yanking. That special category is when the compat bounds have been set too
wide. i.e. say `v2.10.0` was released using a feature not on julia `v1.6` but the compat entry for julia was
not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would not 
solve the issue as on julia v1.6 Pkg would still resolve the broken `v2.10.0`, and as a minor bump, reverting
the code changes would not be valid in a patch bump. In this case one may either submit a PR to retroactively
adjust the compat bounds of previous versions (best user-facing results, but slow and error-prone to implement)

> or yank the offending release.

Reviewer 2: We could also link colprac here, which has some detailed guidance: https://github.com/SciML/ColPrac?tab=readme-ov-file#accidental-breaking-releases and https://github.com/SciML/ColPrac?tab=readme-ov-file#accidental-support-for-an-unsupported-dependency

Author: ```suggestion
If a release is very broken (e.g. contains security vulnerabilities or bugs that are significantly worse 
than the package erroring on load), then it may be yanked. Releasing an ordinary patch release—potentially 
reusing the exact same source code as a previous release without the major bug—is a faster, easier to 
deploy, and less disruptive way to prevent most users from installing a specific version. If you seek to
yank a very broken release, you should typically also release a patch release.

There is however, a special category of bugged releases that can not be resolved by having a patch release.
These also may to be resolved by yanking. That special category is when the compat bounds have been set too
wide. i.e. say `v2.10.0` was released using a feature not on julia `v1.6` but the compat entry for julia was
not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would not 
solve the issue as on julia v1.6 Pkg would still resolve the broken `v2.10.0`, and as a minor bump, reverting
the code changes would not be valid in a patch bump. In this case one may either submit a PR to retroactively
adjust the compat bounds of previous versions (best user-facing results, but slow and error-prone to implement)
or yank the offending release. See the [SciML collaborative practices for more guidance](https://github.com/SciML/ColPrac?tab=readme-ov-file#accidental-support-for-an-unsupported-dependency)
```

If yanking is urgent, open a PR and raise it on the `#pkg-registration` [slack channel](https://julialang.org/slack/)

For releases with actively exploited security vulnerabilities (i.e. malicious code) yanking is not suficient.

> In this case, the release should be completely deleated from the registry and a new patch release should be

Reviewer 2: ```suggestion
In this case, the release should be completely deleted from the registry and a new patch release should be
```

> issued. This is a drastic measure that breaks reporoducivility garuntees and has never been warrented as of 
> April, 2024.

Reviewer 2: ```suggestion
issued. This is a drastic measure that breaks reproducibility guarantees and has never been performed as of 
April, 2024.
```
